src/utils/ref_map.rs

```diff
@@ -0,0 +1,133 @@
+use std::{borrow::Borrow, collections::HashMap, hash::Hash};
+
+/// Trait primarily added to accommodate both native and FFI usage of maps.
+///
+/// In FFI, only object references are available. So, if something like a [`HashMap`] is required,
+/// then it has to be constructed from those references.
+///
+/// In native Rust code, the tendency is to already have the [`HashMap`] containing owned values
+/// so it would be great to be able to work with that rather tha construct a new [`HashMap`] with
+/// references from the one we already have.
+pub trait ReferencesMap<K, V>
+where
+    K: std::hash::Hash + Eq,
+{
+    type Iter<'a>: IntoIterator<Item = (&'a K, &'a V)>
+    where
+        K: 'a,
+        V: 'a,
+        Self: 'a;
+
+    fn get_ref<Q>(&self, key: &Q) -> Option<&V>
+    where
+        for<'a> &'a K: Borrow<Q>,
+        K: Borrow<Q>,
+        Q: Hash + Eq;
+
+    fn iter_ref(&self) -> Self::Iter<'_>;
+}
+
+impl<K, V> ReferencesMap<K, V> for HashMap<K, V>
+where
+    K: std::hash::Hash + Eq,
+{
+    type Iter<'a> = std::collections::hash_map::Iter<'a, K, V>
+    where
+        K: 'a,
+        V: 'a,
+        Self: 'a;
+
+    fn get_ref<Q>(&self, key: &Q) -> Option<&V>
+    where
+        for<'a> &'a K: Borrow<Q>,
+        K: Borrow<Q>,
+        Q: Hash + Eq,
+    {
+        self.get(key)
+    }
+
+    fn iter_ref(&self) -> Self::Iter<'_> {
+        self.iter()
+    }
+}
+
+type MapFnBothRef<'a, K, V> = fn((&'a &'a K, &'a &'a V)) -> (&'a K, &'a V);
```

Comment:
I think it would be cleaner to implement Iterator on a custom struct holding a reference to the HashMap.

Reply:
That's quite the boilerplate for achieving the same thing. I'm sure that dropping some comments with explicit examples on top of these aliases would make things clear enough. I'll give that a shot and I'll wait to see what you think.

Reply:
Forget what I said, you're right 😅 . For some reason I was imagining a lot of boilerplate code as I was thinking of wrapping the `HashMap` into some type which would implement `IntoIterator` and then that also required a dedicate `Iterator` implementor type because you still have to map the values. It felt like a lot more than a type alias.

But since all iterators implement `IntoIterator` it's actually quite succinct. And cleaner. And it has the added benefit that the map closure can be inlined by the compiler, in contrast to the function pointer it was previously using.

Good catch!